src/couch_views/src/couch_views_indexer.erl

```diff
@@ -165,48 +171,60 @@ add_error(Error, Reason, Data) ->
 
 
 update(#{} = Db, Mrst0, State0) ->
-    Limiter = maps:get(limiter, State0),
-    case couch_rate:budget(Limiter) of
-        0 ->
-            couch_rate:wait(Limiter),
-            update(Db, Mrst0, State0);
-        Limit ->
-            {Mrst1, State1} = do_update(Db, Mrst0, State0#{limit => Limit, limiter => Limiter}),
-            case State1 of
-                finished ->
-                    couch_eval:release_map_context(Mrst1#mrst.qserver);
-                _ ->
-                    couch_rate:wait(Limiter),
-                    update(Db, Mrst1, State1)
+    Limit = couch_views_batch:start(),
+    try
+        {Mrst1, State1} = do_update(Db, Mrst0, State0#{limit => Limit}),
+        case State1 of
+            finished ->
+                couch_eval:release_map_context(Mrst1#mrst.qserver);
+            _ ->
+                #{
+                    docs_read := DocsRead,
+                    tx_size := TxSize
+                } = State1,
+                couch_views_batch:success(TxSize, DocsRead),
+                update(Db, Mrst1, State1)
+        end
+    catch
+        error:{erlfdb_error, Error} ->
+            case lists:member(Error, ?RETRY_FAILURES) of
+                true ->
+                    couch_views_batch:failure(),
+                    update(Db, Mrst0, State0);
+                false ->
+                    erlang:error({erlfdb_error, Error})
```

Comment:
Will this loose the original stacktrace? It might not be as nice but we could defined ?RETRY_FAILURES as a guard maybe. Would this work:

```erlang
-define(RETRY_FAILURE(E), (E =:= 1007 orelse E =:= 1031 ...)).
```
And then say `catch error:{erlfdb_error, E} when ?RETRY_FAILURE(E) ...` ?

Reply:
That seems reasonable. I'll update to use that pattern.